portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -40,15 +40,26 @@
   /** Retrieves and outputs JSON based on all user comments */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    
     // Create Query instance for Comment entities and sort by most recent comment first
     Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
 
     // Contains all entities of type Comment in the Datastore 
     PreparedQuery results = datastore.prepare(query);
 
     List<Comment> comments = new ArrayList<>();
+    int numberOfComments = getNumOfComments(request);
+    if (numberOfComments == -1) {
+      // Return empty JSON response
+      response.setContentType("application/json");
+      response.getWriter().println("");
+    }
+
+    int i = 0;
     for (Entity entity : results.asIterable()) {
-      
+      if (i == numberOfComments)
+        break;
+
```

Comment:
An opportunity for optimization: 

This servlet currently queries all the comments, iterates over all of them, and only returns the first numDisplay comments. We've saved a lot on network latency since we will only send numDisplay comments back to the user but we can do even better by limiting the number of comments fetched from the datasore from the outset.

https://cloud.google.com/appengine/docs/standard/java/datastore/retrieving-query-results#setting_limits

If you can use that, you should also be able to get rid of the manual counting with `i`since the query will return at most the number of comments you limit to.